frontend/src/routes/password.change.tsx

```diff
@@ -0,0 +1,293 @@
+    event.preventDefault();
+
+    const formData = new FormData(event.currentTarget);
+
+    const oldPassword = formData.get("current_password");
+    const newPassword = formData.get("new_password");
+    const newPasswordAgain = formData.get("new_password_again");
+
+    if (!oldPassword || !newPassword || !newPasswordAgain) {
+      throw new Error("not all passwords set");
+    }
+
+    // Assert that these are strings — according to the type system they could be Files
+    if (typeof oldPassword !== "string") {
+      throw new Error("oldPassword not string");
+    }
+    if (typeof newPassword !== "string") {
+      throw new Error("newPassword not string");
+    }
+
+    if (newPassword !== newPasswordAgain) {
+      throw new Error("passwords mismatch; this should be checked by the form");
+    }
+
+    changePassword({ userId, oldPassword, newPassword });
+  };
+
+  const success =
+    result.data && result.data.setPassword.status == SetPasswordStatus.Allowed;
+  const handleableError =
+    result.data && result.data.setPassword.status != SetPasswordStatus.Allowed;
+  const unhandleableError = result.error !== undefined;
+
+  const errorMsg: string | undefined = ((): string | undefined => {
+    switch (result.data?.setPassword.status) {
+      case SetPasswordStatus.InvalidNewPassword:
+        return t(
+          "frontend.password_change.failure.description_INVALID_NEW_PASSWORD",
+        );
+      case SetPasswordStatus.NoCurrentPassword:
+        return t(
+          "frontend.password_change.failure.description_NO_CURRENT_PASSWORD",
+        );
+      case SetPasswordStatus.NotAllowed:
+        return t("frontend.password_change.failure.description_NOT_ALLOWED");
+      case SetPasswordStatus.NotFound:
+        return t("frontend.password_change.failure.description_NOT_FOUND");
+      case SetPasswordStatus.PasswordChangesDisabled:
+        return t(
+          "frontend.password_change.failure.description_PASSWORD_CHANGES_DISABLED",
+        );
+      case SetPasswordStatus.WrongPassword:
+        return t("frontend.password_change.failure.description_WRONG_PASSWORD");
+      default:
+        return undefined;
+    }
+  })();
+
+  return (
+    <Layout>
+      <BlockList>
+        <PageHeading
+          Icon={IconLockSolid}
+          title={t("frontend.password_change.title")}
+          subtitle={t("frontend.password_change.subtitle")}
+        />
+
+        <Form.Root onSubmit={onSubmit} method="POST">
+          {/*
+            In normal operation, the submit event should be `preventDefault()`ed.
+            method = POST just prevents sending passwords in the query string,
+            which could be logged, if for some reason the event handler fails.
+          */}
+
+          {!success && (
+            <>
+              <Form.Field
+                name="current_password"
+                serverInvalid={
+                  result.data &&
+                  result.data.setPassword.status ==
```

Comment:
You could shorten this with an optional chaining: 
```suggestion
                  result.data?.setPassword.status ===
```

(also triple equals please :))